### DISCUSSION TITLE
Lazy local variable evaluated once

### DISCUSSION THREAD

**Discussion by emphasis87** [dotnet/csharplang#6995] [Language Ideas]
I would like to have a language syntactic sugar for declaring local variable to be
- evaluated at the call site / lazy
- evaluated only once / cached
- preferably to be readonly

I have read [#312], but I would very much prefer this to focus on local variable implementation only, ideally without addition Lazy<> allocation.

Example usage:
```cs
// "let" or another keyword
let result = await LongEvaluation(); // or a very long expression

if (someCondition)
{
  if (result is string text) // evaluated here
  {
    return result; // already evaluated
  }
}
else if (anotherCondition)
{
  return result + "\n"; // or evaluated here
}

result = ""; // error
```

What possible proposals/discussions are related to this? I remember there being a discussion about local readonly variables, but could not find it.

**EDIT1:**
Perhaps let me rephrase. I don't have a clear answer how to implement this, which is why I would like proposals and discussion, instead of questions to me "how". I am not a compiler designer.

I imagine one solution would be for the compiler to analyze the code flow and create new local variables along certain paths, including inlining the piece of code, so as not to have to track if it is initialized.

**EDIT2**

```cs
let result = Very.Long.Text.I.Would.Rather.Not.Type.Again.But.Not.Sure.If.Evaluation.Is.Needed;

if (condition1)
{
  // here the code could be inlined
  if (result)
  {
    Console.WriteLine("a");
  }
}

if (condition2)
{
  // here could be a local variable + inlined code
  if (result)
  {
    // variable 1 is used again
    Console.WriteLine(result);
  }
}

if (condition2)
{
  // here could be another local variable + inlined code
  if (result)
  {
    // variable 2 is used again
    Console.WriteLine(result);
  }
}
```

However is seems that Lazy<> or LazyInitializer would be needed for lambda capture, or it would have to be restricted.

**Comment by CyrusNajmabadi** [comment]
> ideally without addition Lazy<> allocation.

How would you track/know if it had been initiated then?

**Comment by emphasis87** [comment]
To be clear, it would roughly be equal to:

```cs
// let result = LongEvaluation();

var isInitialized = false;
TResult result;
var resultInit = () => { result = LongEvaluation(); isInitialized = true; return result; } // hopefully inlined within the function

// if (result is string) { }
if ((isInitialized ? result : result = resultInit()) is string) { }
```

**Reply by CyrusNajmabadi** [reply]
Looks like LazyInitializer (https://learn.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer?view=net-7.0). But this would need an object to be thread safe (def needed since this could be captured into lambdas).  So might as well just use a Lazy :-)

**Reply by emphasis87** [reply]
The code is to illustrate the function, not the implementation.

**Reply by CyrusNajmabadi** [reply]
I'm saying the impl would likely have to go this route to be safe.

**Reply by emphasis87** [reply]
Perhaps. I thought that the thread safety is irrelevant for this case.

**Reply by CyrusNajmabadi** [reply]
Seems like it would be important. Imagine a local captured by a lambdas executed on multiple threads. Only one of them would actually produce the result.

**Reply by emphasis87** [reply]
Yeah, that is not what I originally considered. I guess in that scenario you would indeed need Lazy<>. On the other hand in simple scenarios it would not be needed depending on the code flow.

**Reply by emphasis87** [reply]
Any other considerations? If perhaps lambda capture was not allowed...

**Comment by CyrusNajmabadi** [comment]
Generally speaking, I don't like this as it makes it highly unclear what the semantics are. Imagine:

`lazy x = something (y++):`

When does this y++ happen?  At the time the Lazy is created?  Or at the time that you need the value of x. I'd prefer actually using the Lazy type as the semantics would be entirely clear.

**Reply by emphasis87** [reply]
The point is to avoid allocation of Lazy<> by definition as in this example:

```cs
let result = Very.Long.Text.I.Would.Rather.Not.Type.Again.But.Not.Sure.If.Evaluation.Is.Needed;

if (condition1)
{
  // here the code could be inlined
  if (result)
  {
    Console.WriteLine("a");
  }
}

if (condition2)
{
  // here could be a local variable + inlined code
  if (result)
  {
    // variable 1 is used again
    Console.WriteLine(result);
  }
}

if (condition2)
{
  // here could be another local variable + inlined code
  if (result)
  {
    // variable 2 is used again
    Console.WriteLine(result);
  }
}
```

**Reply by emphasis87** [reply]
Perhaps is does not make sense for side effects, I am certainly not sure in all effects.

**Reply by CyrusNajmabadi** [reply]
> The point is to avoid allocation of Lazy<> 

I'm saying that the compiler will likely need to do this anyways to get correct semantics. :-)

**Reply by emphasis87** [reply]
Maybe there could be some low hanging fruit for some cases, where the analysis could be done.

**Comment by DavidArno** [comment]
To answer @CyrusNajmabadi's question of how would this be implemented, that's easy:
1. Step one, the team implement #1051/#832
2. Step 2, the OP writes the following code:
```cs
string SomeMethod()
{
    static string Result()
    {
        static string? lazyValue = null;
        return lazyValue ?? (lazyValue = LongEvaluation());
    }

    if (someCondition)
    {
        if (Result() is string text) // LongEvaluation() is evaluated here
        {
            return Result(); // lazyValue already evaluated so LongEvaluation() isn't called again
        }
    }
}
```
Once `LongEvaluation()` has been called once, subsequent calls to `SomeMethod()` that queried `Result()` would get the cached value.

**Reply by HaloFour** [reply]
That approach doesn't offer any thread safety, not to mention doesn't work if `null` is a legal result from `LongEvaluation`.

My opinion has always been that "lazy" includes too many policy questions to be boiled down to a simple language feature.  I'd rather see it be easier (and more light-weight) to use `Lazy<T>` or `LazyInitializer<T>` in those scenarios where it is useful vs. trying to shoehorn an opinionated approach that, when used wrong, will result in undefined behavior.

**Reply by CyrusNajmabadi** [reply]
We definitely could not implement it that way. For one thing, it would recompute every time the expression evaluated to null.

**Reply by DavidArno** [reply]
> That approach doesn't offer any thread safety,

Details like using a lock just complicate the example. And you missed the point anyway: all that's needed to satisfy the OP is to support static locals and they can write the rest themself. Of course, adding syntactic sugar allowing it all to be expressed as:
```cs
lazy string result = LongEvaluation();
```
and have the compiler generate all the code behind would be icing on the cake, but not essential.

> not to mention doesn't work if `null` is a legal result from `LongEvaluation`.

All the more reason to get unions implemented as the type could then be `string|none`, or somesuch, rather than `string?`

**Reply by DavidArno** [reply]
> We definitely could not implement it that way. For one thing, it would recompute every time the expression evaluated to null.

I've updated the code to clarify that this requirement is easily solved by the language team implementing local statics and the OP implementing the rest.

**Reply by HaloFour** [reply]
> Details like using a lock just complicate the example. And you missed the point anyway: all that's needed to satisfy the OP is to support static locals and they can write the rest themself.

Except that locals can be captured, which makes thread safety a real concern here and with `static` locals.

**Reply by DavidArno** [reply]
> > Details like using a lock just complicate the example. And you missed the point anyway: all that's needed to satisfy the OP is to support static locals and they can write the rest themself.
> 
> Except that locals can be captured, which makes thread safety a real concern here and with `static` locals.

It makes no sense to capture static locals as they are static, not transient. So don't capture static locals.

**Reply by HaloFour** [reply]
Static locals still have lazy initialization behavior which means that thread safety is still important.

**Reply by HaloFour** [reply]
Note that static locals in both C++ and VB are expected to be thread safe.

**Reply by CyrusNajmabadi** [reply]
Indeed.  threadsafety is hugely important here.  normal static fields have htat, so it would be very strange to not have that as a local.  Indeed, it would harm one of the core use cases (moving static fields to locals).

**Reply by DavidArno** [reply]
> Indeed. threadsafety is hugely important here. normal static fields have htat, so it would be very strange to not have that as a local. Indeed, it would harm one of the core use cases (moving static fields to locals).

Static fields are thread-safe during initialisation of the static elements of the class. That is  completely unrelated to this discussion as we are talking about lazy (re)initialization after the class creation is complete.

**Reply by emphasis87** [reply]
Do I understand it correctly, that static local variable has the same value in between separate runs of the function? Because that is not what I am looking for. I am looking for evaluation once per function call. Basically an alias for an expression, with the caveat to not evaluate what was already evaluated before within the current run.
